### PR TITLE
prevent 'attribute_changed?' from returning nil

### LINKED ISSUE #24220: ActiveModel::Dirty #{attribute_name}_changed? returning nil instead of false
### Steps to reproduce
[Here's a gist using the ActiveRecord bug template demonstrating the issue](https://gist.github.com/batter/6947876631e7d72144c9)

### Expected behavior
I would expect that calling `ModelKlass.new.attribute_name_changed?` to always return `true` or `false`. Specifically in this case it should return `false` if the attribute has yet to changed, since it is a method with a question mark appended to the end of it.

### Actual behavior
It returns `nil` in these cases. This is a departure from the behavior in the rest of the Rails 4.2 releases

### System configuration
**Rails version**:
4.2.6

**Ruby version**:
I saw it happen on 2.2.1 and 2.3.0

### PR DESCRIPTION
**Pull request by Sen-Zhang**
Attempt to fix https://github.com/rails/rails/issues/24220

### CODE REVIEW

**Comment by bestori** [comment]
+1

Also, I believe it's cleaner than **.to_bool**ing it.

**Comment by pixeltrix** [comment]
Unfortunately `present?` is simply a negation of `blank?` so doesn't precisely follow Ruby truthy/falsey semantics - e.g both `""` and `[]` are truthy but return `false` for `present?`. Can you please try an alternative, thanks.

**Comment by bestori** [comment]
@pixeltrix do you believe 

> to_bool

 would be more accurate here?

**Comment by Sen-Zhang** [comment]
@pixeltrix just come up with a more comprehensive fix for the root cause of the issue

**Comment by bestori** [comment]
:+1:

File: activerecord/lib/active_record/attribute.rb [hunk 1]
```diff
@@ -93,7 +93,7 @@ def came_from_user?
     end
 
     def has_been_read?
-      defined?(@value)
+      !defined?(@value).nil?
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/attribute.rb]: Why this had to change?

File: activerecord/lib/active_record/attribute.rb [hunk 2]
```diff
@@ -111,7 +111,10 @@ def hash
     protected
 
     attr_reader :original_attribute
-    alias_method :assigned?, :original_attribute
+
+    def assigned?
+      !original_attribute.nil?
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/attribute.rb]: Why this had to change?

File: activerecord/lib/active_record/attribute.rb (hunk 1 above)
Comment (Sen-Zhang) [inline activerecord/lib/active_record/attribute.rb]: prevent the method from returning nil
same reason to the other change

Comment (rafaelfranca) [inline activerecord/lib/active_record/attribute.rb]: But this is semantically diferent. Before it was checking if `@value` was defined, not is it checking if the return of `@value` is not `nil`. I think it is better to use `instance_variable_defined?`

File: activerecord/test/cases/dirty_test.rb [hunk 3]
```diff
@@ -37,7 +37,8 @@ def setup
   def test_attribute_changes
     # New record - no changes.
     pirate = Pirate.new
-    assert !pirate.catchphrase_changed?
+    assert_kind_of FalseClass, pirate.catchphrase_changed?
```
Comment (rafaelfranca) [inline activerecord/test/cases/dirty_test.rb]: `assert_equal false, pirate.catchphrase_changed?`

File: activerecord/lib/active_record/attribute.rb (hunk 1 above)
Comment (Sen-Zhang) [inline activerecord/lib/active_record/attribute.rb]: Good call :+1: ,  have changed to use `instance_variable_defined?`

File: activerecord/test/cases/dirty_test.rb (hunk 3 above)
Comment (Sen-Zhang) [inline activerecord/test/cases/dirty_test.rb]: Fixed

**Comment by rafaelfranca** [comment]
This looks good to me r? @sgrif

File: activerecord/lib/active_record/attribute.rb (hunk 2 above)
Comment (sgrif) [inline activerecord/lib/active_record/attribute.rb]: I agree that this method should not have needed to change. `accessed_fields` isn't documented as returning a boolean, and I don't want to change it to be.

File: activerecord/lib/active_record/attribute.rb [hunk 4]
```diff
@@ -93,7 +93,7 @@ def came_from_user?
     end
 
     def has_been_read?
-      defined?(@value)
+      instance_variable_defined?(:@value)
```
Comment (sgrif) [inline activerecord/lib/active_record/attribute.rb]: This should not need to change.

Comment (rafaelfranca) [inline activerecord/lib/active_record/attribute.rb]: My guess is that the return of `has_been_read?` is being used in the return on `changed?`

Comment (Sen-Zhang) [inline activerecord/lib/active_record/attribute.rb]: `has_been_read?` is being used in the return on `changed_in_place?` and the return of `changed_in_place?` is being used in the return on `changed?`

**Comment by Sen-Zhang** [comment]
just come up with a simpler fix https://github.com/rails/rails/pull/24515

**Comment by sgrif** [comment]
Closing in favor of #24515

**Pull request closed without merging**